# Agent Customization & Prompt Engineering in AutoGen v0.4

In [1]:
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from dotenv import load_dotenv
import os

load_dotenv()
api_key = os.getenv('OPENAI_API_KEY')
model_client = OpenAIChatCompletionClient(model='gpt-4', api_key=api_key)

### Agent Customization

- We can assign a role to our agent
- help in fitting agent to specific use case


In [2]:
asssistant = AssistantAgent(
    name = 'history_expert',
    model_client=model_client,
    description='A knowledgeable assistant with expertise in software engineering',
    system_message='You are a software engineering expert with deep knowledge of software engineering. Provide detailed and accurate answers about software engineering events,design patterns and types of softwares.'
)

In [3]:
async def test_history_expert():
    result = await asssistant.run(task = 'Which design pattern is used in ChatGpt Webchatbot?')
    print(result.messages[-1].content)


In [4]:
await test_history_expert()

OpenAI has not publicly specified the exact design patterns they used in the implementation of ChatGPT. However, given the nature of chatbots like ChatGPT, a few design patterns are common in their implementation:

1. **Model-View-Controller (MVC)**: In this pattern, the application is divided into three interconnected parts. Given the interaction between the user, the AI, and the interface in ChatGPT, it is possible that a pattern like MVC is applied.

2. **Observer Pattern**: In chatbot systems, multiple components need to stay updated with the system’s state. The Observer pattern would allow the state to send updates to these components, maintaining a consistent system.

3. **Strategy Pattern**: AI applications often incorporate multiple algorithms or strategies to solve a problem. The Strategy pattern allows the algorithm to be chosen at runtime, which could potentially be applied in ChatGPT.

4. **Factory Pattern**: For creating different types of responses or processes depending 

OpenAI has not publicly specified the exact design patterns they used in the implementation of ChatGPT. However, given the nature of chatbots like ChatGPT, a few design patterns are common in their implementation:

1. **Model-View-Controller (MVC)**: In this pattern, the application is divided into three interconnected parts. Given the interaction between the user, the AI, and the interface in ChatGPT, it is possible that a pattern like MVC is applied.

2. **Observer Pattern**: In chatbot systems, multiple components need to stay updated with the system’s state. The Observer pattern would allow the state to send updates to these components, maintaining a consistent system.

3. **Strategy Pattern**: AI applications often incorporate multiple algorithms or strategies to solve a problem. The Strategy pattern allows the algorithm to be chosen at runtime, which could potentially be applied in ChatGPT.

4. **Factory Pattern**: For creating different types of responses or processes depending on the context of the user input, the Factory design pattern might be in use.

5. **Singleton Pattern**: Since the conversation with a bot need to be consistent during the session, singleton pattern might be applied to enforce a single instance of conversation context.

Again, it's important to understand that without exact details from the OpenAI team, it's impossible to know the exact paradigm and patterns they used when developing ChatGPT. The above are just examples of patterns that are commonly used in such projects.

# Prompt Engineering

[Prompt Engineering - OpenAI ](https://platform.openai.com/docs/guides/prompt-engineering)

In [5]:
asssistant = AssistantAgent(
    name = 'history_expert',
    model_client=model_client,
    description='A knowledgeable assistant with expertise in world history',
    system_message='You are a history expert with deep knowledge of world history who explaines event in a storytelling style, as if narrating a historic novel. Be vivid and engaging and focus on figures and events'
)

In [6]:
async def test_storytelling():
    result = await asssistant.run(task = 'Tell me about something on the signing of Declaration of Independence')
    print(result.messages[-1].content)

In [7]:
await test_storytelling()

Our tale brings us back to the sweltering summer of 1776, a time that had seen the thirteen American colonies in a bitter and bloody conflict with the mighty British empire. War had been raging for over a year, and the weary colonies, united in their struggle for liberty, were yearning for a means to assert their refusal to remain the subjects of the British crown. 

The city of Philadelphia, the greatest city in the colonies, bore host to the fateful Continental Congress. Imagine the stifling, humid heat; the raucous noises of horses and bustling citizens; the pungent smell of fresh ink and warmed parchment wafting from the Pennsylvania State House, now known as Independence Hall.

It is within these guarded walls that the seeds of a nation were to be sown. Red-faced men, hailing from colonies as far north as New Hampshire and as far South as Georgia, argued passionately amidst a din of flying papers and heated rhetoric. The Congress was divided: moderates led by John Dickinson of Pen

Our tale brings us back to the sweltering summer of 1776, a time that had seen the thirteen American colonies in a bitter and bloody conflict with the mighty British empire. War had been raging for over a year, and the weary colonies, united in their struggle for liberty, were yearning for a means to assert their refusal to remain the subjects of the British crown. 

The city of Philadelphia, the greatest city in the colonies, bore host to the fateful Continental Congress. Imagine the stifling, humid heat; the raucous noises of horses and bustling citizens; the pungent smell of fresh ink and warmed parchment wafting from the Pennsylvania State House, now known as Independence Hall.

It is within these guarded walls that the seeds of a nation were to be sown. Red-faced men, hailing from colonies as far north as New Hampshire and as far South as Georgia, argued passionately amidst a din of flying papers and heated rhetoric. The Congress was divided: moderates led by John Dickinson of Pennsylvania feared the idea of complete independence would breed more strife and suffering; radicals, the likes of John Adams of Massachusetts and Richard Henry Lee of Virginia, sought total severance of the political bonds with the motherland. 

After days of intense debate and bargaining, a draft resolution declaring independence was tentatively accepted on the 2nd of July. The air became electric with possibility and risk, as the men in the room well knew the implications of this daring move; each signature added onto this dangerous document would be considered high treason in the eyes of King George III, a crime punishable by death. 

Enter Thomas Jefferson, the red-haired Virginian polymath, whose soft-spoken demeanor veiled his revolutionary spirit. Having been chosen to pen the Declaration of Independence, Jefferson crafted a timeless and powerful manifesto, justifying the colonies' actions. His evocative prose bore the essence of the Enlightenment era's ideals of liberty, proclaiming the self-evident truths that all men are created equal, with certain unalienable rights that include Life, Liberty, and the pursuit of Happiness. The lines would echo through the chambers of history, equally inspiring and incendiary.

On the 4th of July, the Declaration was officially adopted. As the bells of Philadelphia tolled and spread the news throughout the day, 56 brave delegates stepped forth to sign the document, begining with the President of the Congress, John Hancock. This man's flamboyantly large signature was clearly visible even from a distance, a proud symbol of defiance that would be forever etched into America's birth certificate.

The twilight of 18th century bore witness to a landmark event, the signing of the Declaration of Independence, a monumental act of unity and defiance that marked the dawn of a new nation – The United States of America. This revolutionary event was and remains a dramatic testament to the incredible power of ideas and the indomitable spirit of liberty.

In [16]:
import asyncio
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from dotenv import load_dotenv
import os

# Load API key from .env
load_dotenv()
api_key = os.getenv("OPENAI_API_KEY")
model_client = OpenAIChatCompletionClient(model="gpt-4", api_key=api_key)

# Define a simple tool
def get_weather(city: str) -> str:
    return f"The weather in {city} is sunny!"  # Dummy response

# Set up the agent
agent = AssistantAgent(
    name="weather_agent",
    model_client=model_client,
    system_message="You are a weather assistant. Use the get_weather tool when asked about weather.",
    tools=[get_weather],
    reflect_on_tool_use=True
)

# Run the agent
async def main():
    result = await agent.run(task="What’s the weather in New York?")
    print(result.messages[-1].content)

await main()


The weather in New York is sunny!
